[Mai Ting Kai] iP

tingkai-mai opened the pull request:
### DukeMai ###
A personalised version of Duke, I wrote this for CS2103T's individual project section. This is an event tracker containing:
- **Todo**
- _Deadline_
- ~~Events~~

```
{
  "firstName": "John",
  "lastName": "Smith",
  "age": 25
}
```

It's quite difficult to write this `code`, but I really enjoyed it. I also enjoy:
1. Bread
2. Rice
3. Cake

Things left to do:
- [ ] Update merge request with new commits
- [ ] Check if commits are okay :shipit:
> Have a blockquote

If you've any questions, go to [Google](https://google.com).

Commit:
Add Gradle support

tingkai-mai committed:
Complete first increment which repeats user input

tingkai-mai committed:
Complete second increment which allows saving and printing of user inputs

tingkai-mai committed:
Add Task class

tingkai-mai committed:
Add ability for user to add to a list of tasks where each task can be marked and unmarked

tingkai-mai committed:
Refactor user input logic to use switch statements

tingkai-mai committed:
Refactor toString method to include isComplete status

tingkai-mai committed:
Rename toDoName field to taskTitle

tingkai-mai committed:
Modify Task to initialize isCompleted field to false by default

tingkai-mai committed:
Rename Task taskTitle field to taskName

tingkai-mai committed:
Add Deadline class

tingkai-mai committed:
Add Event class

tingkai-mai committed:
Add ToDo class

tingkai-mai committed:
Allow CLI to handle addition of ToDo, Deadline, Event inputs

tingkai-mai committed:
Rename taskName field to taskDescription

tingkai-mai committed:
Add UnknownCommandException

tingkai-mai committed:
Add MissingDescriptionException

tingkai-mai committed:
Add error handling to Duke

tingkai-mai committed:
Add delete functionality to Duke

tingkai-mai committed:
Add tests for Duke for todo, deadline, event, list, mark, unmark, delete, bye commands

tingkai-mai committed:
Replace user input command with enum COMMANDS

tingkai-mai committed:
Fix bug where invalid user commands cause a runtime error

tingkai-mai committed:
Update tests to include exception handling

tingkai-mai committed:
Add saving of user data to external .txt file

tingkai-mai committed:
Modify EXPECTED.TXT to include file writing operations

tingkai-mai committed:
Add WrongDatetimeFormatException

This exception is thrown when a provided string does not follow a specified datetime format during validation.
A string is passed to this Exception that specifies the correct datetime format to follow, which is printed.

tingkai-mai committed:
Change timeRange field to startDateTime and endDateTime field

Event class now stores the start and end of the event as LocalDateTime objects.

toString() method is also modified to
reflect these changes.

tingkai-mai committed:
Modify by field to store LocalDateTime type

Deadline class now stores the deadline as LocalDateTime object.

toString() method is also modified to
reflect these changes.

tingkai-mai committed:
Modify Duke to convert deadline and event time ranges to LocalDatetime objects

Previously, Duke saves the deadline and event time
ranges as String objects to Deadline and Event objects.

Now, Duke validates and parses user inputs to these
datetimes to LocalDateTime objects and saves them to
Deadline and Event objects.

tingkai-mai committed:
Merge branch 'branch-Level-7'

tingkai-mai committed:
Merge branch 'branch-Level-8'

tingkai-mai committed:
Fix formatting

tingkai-mai committed:
Add relevant header comments to all classes and methods

tingkai-mai committed:
Add DukeException

Represents custom-defined exceptions that are thrown within Duke

tingkai-mai committed:
Modify custom exceptions to extend from DukeException

tingkai-mai committed:
Add Ui class and relevant print statements

tingkai-mai committed:
Convert Ui class methods to static methods

tingkai-mai committed:
Add Storage class and associated methods

tingkai-mai committed:
Fix wrong use of TaskList.size()

tingkai-mai committed:
Add TaskList class and associated methods

tingkai-mai committed:
Add Command class and stubs for Command subclasses

tingkai-mai committed:
Add getEncodedValue function

Returns the encoded string representing a Task object to be saved in the storage

tingkai-mai committed:
Add abstract Command class

Represents an abstract Command that is executed later. Should be inherited and used in subclasses.

tingkai-mai committed:
Create methods for DeadlineCommand representing addition of deadline to task list

tingkai-mai committed:
Create methods for DeleteCommand representing deletion of a task

tingkai-mai committed:
Create methods for EventCommand representing addition of event to tasklist

tingkai-mai committed:
Create methods for ToDoCommand representing addition of Todo to task list

tingkai-mai committed:
Create methods for Goodbye representing exit of Duke

tingkai-mai committed:
Create methods for Goodbye representing an invalid command

tingkai-mai committed:
Create methods for Goodbye representing list command

tingkai-mai committed:
Create methods for MarkCommand representing marking a list item

tingkai-mai committed:
Create methods for UnmarkCommand representing unmarking a list item

tingkai-mai committed:
Make Ui methods static

tingkai-mai committed:
Create methods for TaskList representing the list of tasks

tingkai-mai committed:
Move Storage to commons Package

tingkai-mai committed:
Create Parser class to parse user commands from string

tingkai-mai committed:
Create IllegalIndexException class

Represents illegal access of index or illegal index data type

tingkai-mai committed:
Create IllegalUseException class

Represents when a method in the parent class should not be used directly and used in its subclasses

tingkai-mai committed:
Refactor Duke class to heavily use OOP-based approach

tingkai-mai committed:
Merge branch 'add-gradle-support'

tingkai-mai committed:
Add build.gradle config

tingkai-mai committed:
Rename packages to lowercase and perform refactoring

tingkai-mai committed:
Add checkstyle plugin to gradle build

tingkai-mai committed:
Add dummy jUnit test

tingkai-mai committed:
Write tests for Parser class

Tests for isValidDatetime and splitArrayIntoSubstrings function

tingkai-mai committed:
Modify printSaving to use print()

tingkai-mai committed:
Write tests for Ui class

Test for printSaving function

tingkai-mai committed:
Add JavaDocs to all classes except Command classes and unit test classes

tingkai-mai committed:
Fix formatting

tingkai-mai committed:
Add InsufficientArgumentsException Exception

Thrown when there is insufficient arguments to a function.

tingkai-mai committed:
Add printFindResults function

Prints the results of calling the Find command

tingkai-mai committed:
Add FIND enum and case for FIND

tingkai-mai committed:
Add FindCommand representing case where user wants to find a Task in the current task list

tingkai-mai committed:
Add additional descriptions to runLoop function

tingkai-mai committed:
Add documentation to abstract Command class

tingkai-mai committed:
Add additional newline to for loop as test commit

tingkai-mai committed:
Merge branch 'branch-Level-9'

tingkai-mai committed:
Merge branch 'branch-A-JavaDoc'

tingkai-mai committed:
Merge branch 'branch-A-CodingStandard'

ChongCheeKaiClarence commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,31 @@
+ * provide behavior expected of a Command class.
+ */
+abstract public class Command {
+    protected boolean exit = false;
```
Should the boolean be called exit? Maybe you could use hasExit instead?
```suggestion
    protected boolean hasExit = false;
```

ChongCheeKaiClarence commented on src/main/java/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    public static LocalDateTime parseDeadlineDatetime(String s) {
```
Maybe you could use dateTime as a variable name instead of s? I think it would be easier to keep track of what our variable is.

ChongCheeKaiClarence commented on src/main/java/commands/EventCommand.java:
```diff
@@ -0,0 +1,64 @@
+        String unparsedStartDateTime = unparsedTimeRange.split("-")[0].strip();
+        String unparsedEndDateTime = unparsedTimeRange.split("-")[1].strip();
+
+        if (!Parser.isValidDatetime(unparsedStartDateTime, EVENT_DATETIME_START_OR_END_FORMAT) || !Parser.isValidDatetime(unparsedEndDateTime, EVENT_DATETIME_START_OR_END_FORMAT)) {
```
Does this line feel too long? I think you should have placed a line break.
```suggestion
        if (!Parser.isValidDatetime(unparsedStartDateTime, EVENT_DATETIME_START_OR_END_FORMAT) 
                || !Parser.isValidDatetime(unparsedEndDateTime, EVENT_DATETIME_START_OR_END_FORMAT)) {
```

rexong replied:
This line seems a little to long. Perhaps you can consider the following code:
```suggestion
        bool isValidStartDatetime = !Parser.isValidDatetime(unparsedStartDateTime, EVENT_DATETIME_START_OR_END_FORMAT);
        bool isValidEndDatetime = !Parser.isValidDatetime(unparsedEndDateTime, EVENT_DATETIME_START_OR_END_FORMAT);
        if (!isValidStartDatetime || !isValidEndDatetime) {
```

ChongCheeKaiClarence commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,81 @@
+ * on user-defined tasks. These are saved to an external storage (txt file).
+ */
+public class Duke {
+    static final String storageName = "storage.txt";
```
Should the access modifier be left default?

ChongCheeKaiClarence commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+ * Inherits from Tasks.Task.
+ */
+public class Deadline extends Task {
+    static final String DEADLINE_DATETIME_FORMAT = "d/MM/uuuu HHmm";
```
Should the access modifier be left default?

ChongCheeKaiClarence commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+public class Deadline extends Task {
+    static final String DEADLINE_DATETIME_FORMAT = "d/MM/uuuu HHmm";
+
+    protected LocalDateTime by;
```
Should the access modifier be protected?

ChongCheeKaiClarence commented on src/main/java/tasks/Event.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    @Override
+    public String getEncodedValue() {
+        String formattedStartDateTime = this.startDateTime.format(DateTimeFormatter.ofPattern("MMM dd uuuu, HHmm").withResolverStyle(ResolverStyle.STRICT));
```
Should the code be this long? Consider using a line break.
```suggestion
        String formattedStartDateTime = this.startDateTime
                .format(DateTimeFormatter.ofPattern("MMM dd uuuu, HHmm")
                        .withResolverStyle(ResolverStyle.STRICT));
```

ChongCheeKaiClarence commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @return Boolean representing if a Task is complete.
+     */
+    public boolean getIsComplete() {
```
Personally, I am not sure if the method name is acceptable. I get that you are using a getter method, but for a boolean, shouldn't the name just be IsComplete()?

ChongCheeKaiClarence commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,63 @@
+     *
+     * @param completedStatus Boolean representing if the Task is complete.
+     */
+    public void setIsComplete(boolean completedStatus) {
```
Maybe u could change the method name to updateCompletionStatus()? I think it sounds nicer.
Or maybe you could create a separate method that changes isComplete to true and another method that changes isComplete to false.

ChongCheeKaiClarence commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,61 @@
+     */
+    public Deadline(String taskDescription, boolean isComplete, LocalDateTime by) {
+        super(taskDescription);
+        this.setIsComplete(isComplete);
```
Rather than creating a task with its isComplete = false, wouldn't it be better to make all tasks created to have their IsComplete always initially false?

ChongCheeKaiClarence reviewed (commented):
Looks good. 👍
Very good use of OOP and abstraction.
Some styling changes are requested, especially line breaks.

tingkai-mai committed:
Add JavaFX dependencies to build.gradle file

tingkai-mai committed:
Add MainWindow.fxml

tingkai-mai committed:
Add MainWindow class

tingkai-mai committed:
Add DialogBox.fxml

tingkai-mai committed:
Add DialogBox class

tingkai-mai committed:
Add Main class

Provides start function to initialize JavaFX GUI

tingkai-mai committed:
Add Launcher class

Driver class that launches JavaFX GUI.

rexong commented on src/main/java/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,58 @@
+public class DeadlineCommand extends Command {
+    static final String DEADLINE_DATETIME_FORMAT = "d/MM/uuuu HHmm";
+    static final String DEADLINE_STORAGE_FORMAT = "MMM dd uuuu, HHmm";
+    private final String[] args;
```
Perhaps the name here can be more intuitive? Like maybe you can use `userInput` instead of `args`. Also, I'm not too sure about the convention here, but I think the keyword `final` implies a constant and thus, `args` here should be `ARGS`. What do you think?

rexong commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,36 @@
+import tasklist.TaskList;
+
+public class FindCommand extends Command {
+    String[] args;
```
I think you might have forgotten an acessor here.

rexong commented on src/main/java/common/Parser.java:
```diff
@@ -0,0 +1,148 @@
+     * @param str String to check if the enum exists.
+     * @return True if the provided string is a valid common.COMMANDS enum.
+     */
+    public static boolean validCommandEnum(String str) {
```
Perhaps you can consider naming this method `isValidCommandEnum` instead?

rexong commented on src/main/java/common/Storage.java:
```diff
@@ -0,0 +1,105 @@
+                    break;
+                }
+                case "[E]": {
+                    taskList.addTask(new Event(encodedTask[1], Boolean.parseBoolean(encodedTask[2]), EventCommand.parseEventDatetimeFromStorage(encodedTask[3]), EventCommand.parseEventDatetimeFromStorage(encodedTask[4])));
```
This line here seems a bit too long, perhaps a line break?
```suggestion
                    taskList.addTask(new Event(encodedTask[1], 
                         Boolean.parseBoolean(encodedTask[2]), 
                         EventCommand.parseEventDatetimeFromStorage(encodedTask[3]), 
                         EventCommand.parseEventDatetimeFromStorage(encodedTask[4])));
```

rexong commented on src/test/java/duke/DukeTest.java:
```diff
@@ -0,0 +1,17 @@
+
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
I think static import should be imported before other imports.

rexong reviewed (commented):
LGTM, enjoyed reading your code. Just need to do some minor changes. 😃

tingkai-mai committed:
Depreciate Ui class

Replaced by ChatResponse class as now we print to Duke GUI, not CLI.

tingkai-mai committed:
Add ChatResponse class

Replaces Ui class as now we print to Duke GUI, not CLI.

tingkai-mai committed:
Modify Command classes to return string responses

This is to display their responses via the JavaFX GUI

tingkai-mai committed:
Modify Storage class to return string responses

This is to display its responses via the JavaFX GUI

tingkai-mai committed:
Fix formatting

tingkai-mai committed:
Modify Duke to return Command string responses via getResponse()

Depreciates other functions in Duke that are related to CLI

tingkai-mai committed:
Update ChatResponse documentation

tingkai-mai committed:
Fix Duke initialize message not being displayed on GUI

tingkai-mai committed:
Add DukeTest file

tingkai-mai committed:
Add storage file with seed data

tingkai-mai committed:
Add chat sender and receiver profile pictures

tingkai-mai committed:
Add assertion condition for unparsed datetime string in Deadline

This ensures that the datetime string, obtained from parsing the user input, is not empty.
If it is empty, then Parser validation has failed to detect this.

This prevents empty deadline datetime from being saved to storage.

tingkai-mai committed:
Add assertion condition for unparsed datetime strings in Events

This ensures that the datetime string, obtained from parsing the user input,
is not empty. If it is empty, then Parser validation has failed to detect this.

This prevents empty Event start, end datetimes and date range from being saved
to storage.

tingkai-mai committed:
Add assertion condition for mark command

This ensures that mark command is not called with empty index.

tingkai-mai committed:
Add assertion condition for unparsed todo item string in ToDo

This ensures that an empty Todo is not saved to storage.

tingkai-mai committed:
Merge pull request #1 from Ferusel/branch-A-Assertions

Branch A Assertions

tingkai-mai committed:
Fix formatting in switch statement

tingkai-mai committed:
Add splitString function and fix formatting in switch statement

tingkai-mai committed:
Apply SLAP principle to execute function

tingkai-mai committed:
Merge pull request #2 from Ferusel/branch-A-CodeQuality

Branch A-CodeQuality

tingkai-mai committed:
Rename validCommandEnum to isValidCommandEnum

tingkai-mai committed:
Fix seed data containing wrong labels

tingkai-mai committed:
Add StatisticsCommand Command

Represents new statistics command that
currently allows to obtain the count of
tasks within the task list.

tingkai-mai committed:
Add returnStatisticTaskCount Function

Returns formatted string containing count
of a specified task.

tingkai-mai committed:
Fix assertion statements with inverse boolean logic

tingkai-mai committed:
Add .yml file specifying Github Actions for CI

tingkai-mai committed:
Modify profile picture circular

tingkai-mai committed:
Reduce circular profile picture radius

tingkai-mai committed:
Style DialogBox

tingkai-mai committed:
Style MainWindow

tingkai-mai committed:
Modify ChatResponse to use SotongBoi flavour text

tingkai-mai committed:
Add additional guard clauses checking for missing delimiter in DeadlineCommand

tingkai-mai committed:
Add additional guard clauses checking for missing delimiter in EventCommand

tingkai-mai committed:
Add additional warning in IllegalIndexException error message

tingkai-mai committed:
Add assert statements for validateArguments

Asserts that args is not empty in validateArguments

tingkai-mai committed:
Fix printing of error message

No longer prints the error message class

tingkai-mai committed:
Merge branch 'branch-A-Personality'

tingkai-mai committed:
Merge branch 'branch-A-MoreErrorHandling'

tingkai-mai committed:
Add first draft of README.md

tingkai-mai committed:
Fix bug where empty statistic arguments cause IndexOutOfBounds error

tingkai-mai committed:
Write User Guide

tingkai-mai committed:
Replace DaDuke.png and DaUser.png with SotongBoi.png and User.png

tingkai-mai committed:
Add title to GUI window

tingkai-mai committed:
Add Ui snapshot

tingkai-mai committed:
Update dummy storage data

tingkai-mai committed:
Rename exit to isExit

tingkai-mai committed:
Fix checkstyles

tingkai-mai committed:
Modify build.gradle to build Launcher

tingkai-mai committed:
Fix README not displaying Ui.png properly

tingkai-mai committed:
Fix checkstyle errors

tingkai-mai committed:
Fix java.lang.UnsatisfiedLinkError

Occurs when program runs on Macs with the following error:
Loading library prism_es2 from resource failed: java.lang.UnsatisfiedLinkError: Can't load library

tingkai-mai committed:
Modify javadocs headers in Command and its subclasses
to follow Java coding conventions

tingkai-mai committed:
Modify javadocs headers to follow Java coding conventions

tingkai-mai committed:
Add acknowledgements to README.md, Main, MainWindow classes
